keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2861 @@
+    kind: ReplicaLimit
+    spec:
+      maxReplicas: 100
+```
+
+Pros:
+
+- Matching criteria is fully defined and validated in a builtin type
+
+Cons:
+
+- Embedded config needs the same treatment as custom resource, but
+  reimplementing it all on an embedded resource is, at best, highly impractical
+  in the apiserver as it exists today. E.g.. there is not automatic validation
+  of the embedded resource.
+
+###### Alternative: Generate CRDs
+
+The `ValidatingAdmissionPolicy` contains a OpenAPIv3 schema defining how the
+policy is configured. The schema need only contains the policy specific
+configuration (it does not need to contain match rules or anything else that is
+standard).
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "validate-xyz.example.com"
+spec:
+  config:
+    group: rules.example.com
+    kind: ReplicaLimit
+    version: v1
+    openAPIV3Schema:
+      type: object
+      properties:
+        spec:
+          type: object
+          properties:
+            maxReplicas:
+              type: int
+```
+
+This allows the apiserver to combine the configuration schema with the match
+schema and then generate a CRD. The apiserver then can run a control loop to
+always keep the CRD match stanzas in sync with what is expected.
+
+This could alternatively be kept separate from the policy definition by having
+something like:
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionConfigurationDefinition
+metadata:
+  name: "validate-configuration-xyz.example.com"
+spec:
+  configSchema:
+    group: rules.example.com
+    kind: ReplicaLimit
+    version: v1
+    openAPIV3Schema:
+      type: object
+      properties:
+        spec:
+          type: object
+          properties:
+            maxReplicas:
+              type: int
+```
+
+Pros:
+
+- Policy author doesn't need to define the CRD. It can instead be generated for
+  them.
+
+Cons:
+
+- Most of the cons of "Duck Typed CRDs" alternative apply, since ultimately a
+  CRD is created and behaves the same.
+- This implies that the policy owns the configuration. But we have use cases
+  where multiple policies share the same configuration. For this model it seems
+  misleading and potentially problematic to have both policies attempting to
+  define the same configuration.
+
+##### Match Criteria
+
+During admission, the Kubernetes API server will validate the resource being
+admitted against all policy configuration resources that match the resource.
+
+While webhook match rules give a good sense of what types of capabilities might
+be needed, they serve a slightly different purpose.  Webhook match rules make it
+possible to avoid webhook requests, which incur latency and impact availability,
+for resources that don't need to be evaluated by the webhook. CEL expressions
+have a comparatively small impact of latency and are in-process (and so do not
```

Comment:
How many match criteria will you have to evaluate per mutating change?

Should probably set a max of some sort. That will add up eventually.

(we can compile a decision tree if we eventually need too many)

Reply:
I've added a section to spell out all the limits we will put in place. Ones that come to mind are:

- Max policy bindings per definition
- Max lengths for all lists in match criteria (resourceRules, namespaceSelectors, labelSelectors, ...)
- per expression CEL evaluation cost limits (we will reuse the resource limits we introduced in CRD validation)
- per policy CEL evaluation cost limits